src/agent/types.ts

```diff
@@ -1,14 +1,24 @@
-import type { ServerSurfaceCall } from '@grpc/grpc-js/build/src/server-call';
-import type { Class } from '@matrixai/errors';
-import type { ConnectionInfo } from '../network/types';
-import type ErrorPolykey from '../ErrorPolykey';
+import type { JSONValue } from '../types';
 
-type ConnectionInfoGet = (
-  call: ServerSurfaceCall,
-) => ConnectionInfo | undefined;
+// eslint-disable-next-line
+type NoData = {};
```

Comment:
Using `{}` as a type is valid but is caught by the linter. The `NoData` is used when a RPC message is intended to be empty. But we don't want to use  `// eslint-disable-next-line` each time we want to use an empty message. So we use `NoData` instead.

Reply:
What's the eslint error? Makes me think there should be a type for `{}` empty object already.

Reply:
```/home/faulkes/matrixcode/polykey/js-polykey/src/client/types.ts
  3:15  error  Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead.
- If you want a type meaning "empty object", you probably want `Record<string, never>` instead  @typescript-eslint/ban-types
```

it mentions in the linter docs that the only valid use of `{}` is when you specifically mean an empty object. But there is no way to change the rule to allow this case.

`Record<string, never>` causes issues when combining this type with others.

Reply:
What about https://stackoverflow.com/a/64568494/582917?

Also what do you mean by "combining this type"?

Also just to be precise this returned result is embedded within the `result` property of the JSONRPC message right?

```
{"jsonrpc": "2.0", "result": 19, "id": 1}
```

According to the spec, `null` is allowed. Because `undefined` just means the entire property gets removed. But if you specify `null` it is allowed. Perhaps we should instead use `null`?

```
> JSON.stringify(undefined)
undefined
> JSON.stringify({a: undefined })
'{}'
> JSON.stringify({a: null })
'{"a":null}'
```

Reply:
![image](https://github.com/MatrixAI/Polykey/assets/640797/1cdd0ed6-8f0d-4755-a28c-d2aa90be5863)

Sounds like `null` would be the better thing to return if there's truly nothing to return. The equivalent of "void" in our RPC calls basically.

Reply:
Whatever we return we need to union with whatever the middleware is working with. In this case its the `ClientRPCRequestParams` , `ClientRPCResponseResult`, `AgentRPCRequestParams` and `ClientRPCResponseResult`.

Because we need to union our messages with that structure, it has to be an empty object `{}` so we can spread it with the metadata.

```
type AgentRPCRequestParams<T extends Record<string, JSONValue> = NoData> = {
  metadata?: {
    [Key: string]: JSONValue;
  } & Partial<{
    authorization: string;
    timeout: number;
  }>;
} & Omit<T, 'metadata'>;
```

We'd have to change this structure to allow for `null` to be a valid message.

Reply:
We settled on keeping it as is, but we're centralised the type to `src/types` under `ObjectEmpty`.